platform/mellanox/mlnx-platform-api/sonic_platform/chassis.py

```diff
@@ -95,5 +135,170 @@ def _get_port_position_tuple_by_sku_name(self):
+
+        Returns:
+            A string containing the MAC address in the format
+            'XX:XX:XX:XX:XX:XX'
+        """
+        return self.eeprom.get_base_mac()
+
+    def get_serial_number(self):
+        """
+        Retrieves the hardware serial number for the chassis
+
+        Returns:
+            A string containing the hardware serial number for this chassis.
+        """
+        return self.eeprom.get_serial_number()
+
+    def get_system_eeprom_info(self):
+        """
+        Retrieves the full content of system EEPROM information for the chassis
+
+        Returns:
+            A dictionary where keys are the type code defined in
+            OCP ONIE TlvInfo EEPROM format and values are their corresponding
+            values.
+        """
+        return self.eeprom.get_system_eeprom_info()
+
+    def _read_generic_file(self, filename, len):
+        """
+        Read a generic file, returns the contents of the file
+        """
+        result = ''
+        try:
+            fileobj = io.open(filename)
+            result = fileobj.read(len)
+            fileobj.close()
+            return result
+        except:
+            log_warning("Fail to read file {}, maybe it doesn't exist".format(filename))
+            return ''
+
+    def _verify_reboot_cause(self, filename):
+        '''
+        Open and read the reboot cause file in 
+        /var/run/hwmanagement/system (which is defined as REBOOT_CAUSE_ROOT)
+        If a reboot cause file doesn't exists, returns '0'.
+        '''
+        return bool(int(self._read_generic_file(join(REBOOT_CAUSE_ROOT, filename), REBOOT_CAUSE_FILE_LENGTH).rstrip('\n')))
+
+    def get_reboot_cause(self):
+        """
+        Retrieves the cause of the previous reboot
+
+        Returns:
+            A tuple (string, string) where the first element is a string
+            containing the cause of the previous reboot. This string must be
+            one of the predefined strings in this class. If the first string
+            is "REBOOT_CAUSE_HARDWARE_OTHER", the second string can be used
+            to pass a description of the reboot cause.
+        """
+        #read reboot causes files in the following order
+        minor_cause = ''
+        if self._verify_reboot_cause(REBOOT_CAUSE_POWER_LOSS_FILE):
+            major_cause = self.REBOOT_CAUSE_POWER_LOSS
+        elif self._verify_reboot_cause(REBOOT_CAUSE_THERMAL_OVERLOAD_ASIC_FILE):
+            major_cause = self.REBOOT_CAUSE_THERMAL_OVERLOAD_ASIC
+        elif self._verify_reboot_cause(REBOOT_CAUSE_WATCHDOG_FILE):
+            major_cause = self.REBOOT_CAUSE_WATCHDOG
+        else:
+            major_cause = self.REBOOT_CAUSE_HARDWARE_OTHER
+            if self._verify_reboot_cause(REBOOT_CAUSE_MLNX_FIRMWARE_RESET):
+                minor_cause = "Reset by ASIC firmware"
+            else:
+                reboot_by_user = self._read_generic_file(REBOOT_CAUSE_FILE_USER_REBOOT, REBOOT_CAUSE_USER_REBOOT_MAX_LENGTH)
```

Comment:
You don't need to read the reboot cause file here. The `get_reboot_cause()` function only needs to return the hardware-related reboot cause or `REBOOT_CAUSE_NON_HARDWARE` if the previous reboot was not caused by hardware.

I am moving the reboot-cause processing logic out of rc.local and into a one-shot service which will run at boot. It will call the `get_reboot_cause()` function, and if it returns anything other than `REBOOT_CAUSE_NON_HARDWARE`, it will use that as the reboot reason. If the function returns `REBOOT_CAUSE_NON_HARDWARE`, then the service will use the cause found in /host/reboot-cause/previous-reboot-cause.txt.

Reply:
It's ok to move reboot cause logic out of the pmon docker and handle it on the host. But this requires class Chassis to be loaded on host. 
I just wonder whether it is an issue if some vendor's chassis implementation imports some base classes defined in modules which are installed in pmon docker only, like sonic_platform_daemon, sonic_platform_common or other components required to perform SFP accessing.

Reply:
fixed.

Reply:
>It's better to move reboot cause logic out of the pmon docker and handle it on the host. But this requires class Chassis to be loaded on host.

Yes, the plan is to have the vendor's "sonic_platform" package installed both in the PMon container as well as the host system. We plan on processing reboot cause and managing the watchdog from the host system; all other functionality (for now, at least) will be managed inside the host container.

FYI, I have raised a PR for the new "process-reboot-cause" service, if you'd like to take a look: https://github.com/Azure/sonic-buildimage/pull/3102